File: 2020/23b.py
class Cup(object):
	def __init__(self, label):
		self.label = label

def make_move(current_cup, cup_map):
	num_cups = len(cup_map)
	first_selected = current_cup.next
	middle_selected = first_selected.next
	last_selected = middle_selected.next
	new_current_cup = last_selected.next

	selected_labels = {first_selected.label, middle_selected.label, last_selected.label}
	destination_label = (current_cup.label - 2) % num_cups + 1
	while destination_label in selected_labels:
		destination_label = (destination_label - 2) % num_cups + 1
	destination_cup = cup_map[destination_label - 1]
	after_destination_cup = destination_cup.next

	destination_cup.next = first_selected
	last_selected.next = after_destination_cup
	current_cup.next = new_current_cup
	return new_current_cup

# create cups
cup_map = [Cup(x + 1) for x in range(1000000)]

# read given cups
given_cups = [cup_map[int(x) - 1] for x in input()]

# connect given cups
for i in range(1, len(given_cups)):
	given_cups[i - 1].next = given_cups[i]

# remaining cups
for i in range(len(given_cups) + 1, 1000000):
	cup_map[i - 1].next = cup_map[i]

#connect given cups with the rest
given_cups[-1].next = cup_map[len(given_cups)]
cup_map[-1].next = given_cups[0]

current_cup = given_cups[0]
for i in range(10000000):
	current_cup = make_move(current_cup, cup_map)

print(cup_map[0].next.label * cup_map[0].next.next.label)

File: 2020/23b.rs
use std::io;

const NUM_CUPS: usize = 1_000_000;

fn make_move(nexts: &mut [u32], current: u32) -> u32 {
	let first_selected = nexts[current as usize];
	let middle_selected = nexts[first_selected as usize];
	let last_selected = nexts[middle_selected as usize];
	let new_current = nexts[last_selected as usize];

	let selected = [ first_selected, middle_selected, last_selected ];
	let mut destination = (current + NUM_CUPS as u32 - 1) % NUM_CUPS as u32;
	while selected.contains(&destination) {
		destination = (destination + NUM_CUPS as u32 - 1) % NUM_CUPS as u32;
	}
	let after_destination = nexts[destination as usize];

	nexts[destination as usize] = first_selected;
	nexts[last_selected as usize] = after_destination;
	nexts[current as usize] = new_current;

	new_current
}

fn main() {
	// create connections
	let mut nexts: Vec<u32> = (0..NUM_CUPS as u32).map(|i| i + 1 % NUM_CUPS as u32).collect();

	// change connections for given cups
	let mut line = String::new();
	io::stdin().read_line(&mut line).unwrap();
	let given_cups = line.trim();
	let mut prev_label = NUM_CUPS as u32 - 1;
	for c in given_cups.chars() {
		let current_label = c.to_digit(10).unwrap() - 1;
		nexts[prev_label as usize] = current_label;
		prev_label = current_label;
	}
	nexts[prev_label as usize] = given_cups.len() as u32;

	// execute moves
	let mut current = nexts[NUM_CUPS - 1];
	for _ in 0..10_000_000 {
		current = make_move(nexts.as_mut(), current);
	}

	println!("{}", (nexts[0] as u64 + 1) * (nexts[nexts[0] as usize] as u64 + 1));
}
